README.rst
A small library containing high level abstractions for driving Galaxy via Selenium.

* Free software: Academic Free License version 3.0
* Documentation: https://galaxy-selenium.readthedocs.org.
* Code: https://github.com/galaxyproject/galaxy-selenium

An Example
----------

The following example demonstrates the usage of this library by automating a few
tasks in Galaxy. This example will register a new user, upload this README via the
tool upload form, open the "Random Lines" tool, and execute it on the uploaded file.

::

    import os

    from random import randint

    from galaxy_selenium.cli import DriverWrapper

    driver_wrapper = DriverWrapper(galaxy_url="http://localhost:8080")

    driver_wrapper.register("mary%s@galaxyproject.org" % randint(0, 100000))
    driver_wrapper.perform_upload(os.path.abspath("./README.rst"))
    driver_wrapper.history_panel_wait_for_hid_ok(1)

    driver_wrapper.click_label("Text Manipulation")
    driver_wrapper.tool_open("random_lines1")
    driver_wrapper.tool_set_value("num_lines", "5")
    driver_wrapper.tool_execute()
    driver_wrapper.history_panel_wait_for_hid_ok(2)

    driver_wrapper.history_panel_click_item_title(hid=2)

    driver_wrapper.driver.save_screenshot("./randomlines.png")


HISTORY.rst
.. :changelog:

History
-------

.. to_doc

---------------------
17.9.0.dev0
---------------------

* Setup project.

.. _bioblend: https://github.com/galaxyproject/bioblend/
.. _XSD: http://www.w3schools.com/schema/
.. _lxml: http://lxml.de/
.. _xmllint: http://xmlsoft.org/xmllint.html
.. _nose: https://nose.readthedocs.org/en/latest/


sync_galaxy_root.sh
#!/bin/bash

usage() {
cat << EOF
Usage: ${0##*/} [-i] /path/to/galaxy...
Sync galaxy-selenium modules to those same modules in Galaxy directory (or vice versa if -i).

EOF
}

if [ $# -lt 1 ]; then
    usage
    exit 1
fi

invert=0
OPTIND=1
while getopts ":i" opt; do
    case "$opt" in
        h)
            usage
            exit 0
            ;;
        i)
            invert=1
            ;;
        '?')
            usage >&2
            exit 1
            ;;
    esac
done
shift "$((OPTIND-1))" # Shift off the options and optional --.

PROJECT_DIRECTORY="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
GALAXY_DIRECTORY=$1



if [ "$invert" -ne "1" ];
then

    rm -rf $GALAXY_DIRECTORY/test/galaxy_selenium
    cp -r $PROJECT_DIRECTORY/test/galaxy_selenium $GALAXY_TEST_DIR/test/galaxy_selenium
    
else

    rm -rf $PROJECT_DIRECTORY/galaxy_selenium/*
    cp -r $GALAXY_LDIRECTORY/test/galaxy_selenium $PROJECT_DIRECTORY/

fi


galaxy_selenium/sizzle.py
"""Utilities for using sizzle (jQuery-style) selectors with Selenium."""

import re

from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.support.ui import WebDriverWait

from .has_driver import exception_indicates_stale_element

SIZZLE_LOAD_TIMEOUT = 15
SIZZLE_URL = "//cdnjs.cloudflare.com/ajax/libs/sizzle/1.10.18/sizzle.js"


def sizzle_selector_clickable(selector):
    def ec(driver):
        elements = find_elements_by_sizzle(driver, selector)
        if not elements:
            return False
        element = elements[0]
        try:
            clickable = element.is_displayed() and element.is_enabled()
        except Exception as e:
            # Handle the case where the element is detached between when it is
            # discovered and when it is checked - it is likely changing quickly
            # and the next pass will be the final state. If not, this should be
            # wrapped in a wait anyway - so no problems there. For other
            # non-custom selectors I believe this all happens on the Selenium
            # server and so there is likely no need to handle this case - they
            # are effectively atomic.
            if exception_indicates_stale_element(e):
                return None

            raise
        if clickable:
            return element
        else:
            return None

    return ec


def sizzle_presence_of_selector(selector):
    def ec(driver):
        elements = find_elements_by_sizzle(driver, selector)
        if not elements:
            return False
        element = elements[0]
        try:
            displayed = element.is_displayed()
        except Exception as e:
            # See note above insizzle_selector_clickable about this exception.
            if exception_indicates_stale_element(e):
                return None

            raise

        if displayed:
            return element
        else:
            return None

    return ec


def find_element_by_sizzle(driver, sizzle_selector):
    """
    Finds an element by sizzle selector.
    :Args:
     - sizzle_selector: The sizzle selector to use when finding element.
    :Usage:
        driver.find_element_by_sizzle('#foo')
    """
    elements = driver.find_elements_by_sizzle(sizzle_selector)
    if elements:
        return elements[0]
    else:
        raise NoSuchElementException(
            "Unable to locate element by Sizzle: {selector}".format(selector=sizzle_selector)
        )


def find_elements_by_sizzle(driver, sizzle_selector):
    """
    Finds elements by sizzle selector.
    :Args:
     - sizzle_selector: The sizzle selector to use when finding elements.
    :Usage:
        driver.find_elements_by_sizzle('.foo')
    """
    if not _is_sizzle_loaded(driver):
        _inject_sizzle(driver, SIZZLE_URL, SIZZLE_LOAD_TIMEOUT)
    elements = driver.execute_script(_make_sizzle_string(sizzle_selector))
    return elements


def _inject_sizzle(driver, sizzle_url, timeout):
    script = """
        var _s = document.createElement("script");
        _s.type = "text/javascript";
        _s.src = "{src}";
        var _h = document.getElementsByTagName("head")[0];
        _h.appendChild(_s);
    """.format(src=sizzle_url)
    driver.execute_script(script)
    wait = WebDriverWait(driver, timeout)
    wait.until(lambda d: _is_sizzle_loaded(d),
               "Can't inject Sizzle in {timeout} seconds".format(timeout=timeout))


def _is_sizzle_loaded(driver):
    script = "return typeof(Sizzle) != \"undefined\";"
    return driver.execute_script(script)


def _make_sizzle_string(sizzle_selector):
    try:
        selector = sizzle_selector.decode("utf-8")
    except (AttributeError, UnicodeEncodeError):
        selector = sizzle_selector
    return u"return Sizzle(\"{selector}\");".format(selector=re.escape(selector))


__all__ = (
    "sizzle_selector_clickable",
    "sizzle_presence_of_selector",
    "find_element_by_sizzle",
    "find_elements_by_sizzle",
)


galaxy_selenium/dump_tour.py
#!/usr/bin/env python
import argparse
import datetime
import os
import time
import sys

from galaxy_selenium import cli

DESCRIPTION = "Walk a Galaxy tour and dump screenshots."


def main(argv=None):
    if argv is None:
        argv = sys.argv[1:]

    args = _arg_parser().parse_args(argv)
    driver_wrapper = cli.DriverWrapper(args)
    output = args.output
    assert output is not None
    output = output.replace("TIMESTAMP", datetime.datetime.now().strftime("%Y%m%d%H%M%s"))
    if os.path.exists(output):
        raise Exception("Output directory %s exists, skipping")
    else:
        os.makedirs(output)
    callback = DumpTourCallback(driver_wrapper, output)
    driver_wrapper.run_tour(args.tour, tour_callback=callback)


class DumpTourCallback(object):

    def __init__(self, driver_wrapper, output):
        self.driver_wrapper = driver_wrapper
        self.output = output

    def handle_step(self, step, step_index):
        time.sleep(.5)
        self.driver_wrapper.driver.save_screenshot("%s/%i.png" % (self.output, step_index))
        time.sleep(.5)


def _arg_parser():
    parser = argparse.ArgumentParser(description=DESCRIPTION)
    parser.add_argument('tour', metavar='TOUR', help='tour to walk')
    parser.add_argument('-o', '--output', default="tour_dump_TIMESTAMP", help='directory to dump tour to')
    parser = cli.add_selenium_arguments(parser)
    return parser


if __name__ == "__main__":
    main()


galaxy_selenium/cli.py
REMOTE_DESCRIPTION = "Use a remote driver for selenium connection."
BROWSER_DESCRIPTION = "Use specific browser for selenium (e.g. firefox or chrome)."
REMOTE_HOST_DESCRIPTION = "Selenium hub remote host to use (if remote driver in use)."
REMOTE_PORT_DESCRIPTION = "Selenium hub remote port to use (if remote driver in use)."
GALAXY_URL_DESCRIPTION = "URL of Galaxy instance to target."
HEADLESS_DESCRIPTION = "Use local selenium headlessly (requires pyvirtualdisplay)."

from six.moves.urllib.parse import urljoin

from .driver_factory import (
    get_local_driver,
    get_remote_driver,
    virtual_display_if_enabled,
)
from .navigates_galaxy import NavigatesGalaxy


def add_selenium_arguments(parser):
    """Add common selenium arguments for argparse driver utility."""

    parser.add_argument(
        '--selenium-browser',
        default="auto",
        help=BROWSER_DESCRIPTION,
    )
    parser.add_argument(
        '--selenium-headless',
        default=False,
        action="store_true",
        help=HEADLESS_DESCRIPTION,
    )
    parser.add_argument(
        '--selenium-remote',
        default=False,
        action="store_true",
        help=REMOTE_DESCRIPTION,
    )
    parser.add_argument(
        '--selenium-remote-host',
        default="127.0.0.1",
        help=REMOTE_HOST_DESCRIPTION,
    )
    parser.add_argument(
        '--selenium-remote-port',
        default="4444",
        help=REMOTE_PORT_DESCRIPTION,
    )
    parser.add_argument(
        '--galaxy-url',
        default="http://127.0.0.1:8080/",
        help=GALAXY_URL_DESCRIPTION,
    )

    return parser


def args_to_kwds(args):
    kwds = {
        'selenium_browser': args.selenium_browser,
        'selenium_headless': args.selenium_headless,
        'selenium_remote': args.selenium_remote,
        'host': args.selenium_remote_host,
        'port': args.selenium_remote_port,
        'galaxy_url': args.galaxy_url,
    }
    return kwds


class DriverWrapper(NavigatesGalaxy):

    """Adapt argparse command-line options to a concrete Selenium driver."""

    def __init__(self, args=None, **kwds):
        if args is not None:
            assert not kwds
            kwds = args_to_kwds(args)

        self.target_url = kwds["galaxy_url"]
        # TODO: Allow overriding this.
        self.target_url_from_selenium = self.target_url

        browser = kwds.get("selenium_browser", "auto")
        self.display = virtual_display_if_enabled(kwds.get("selenium_headless", False))
        if kwds.get("selenium_remote", False):
            driver = get_remote_driver(
                host=kwds["selenium_remote_host"],
                port=kwds["selenium_remote_port"],
                browser=browser,
            )
        else:
            driver = get_local_driver(
                browser=browser,
            )
        self.driver = driver

    def build_url(self, url="", for_selenium=False):
        if for_selenium:
            return urljoin(self.target_url_from_selenium, url)
        else:
            return urljoin(self.target_url, url)

    @property
    def default_timeout(self):
        return 15

    def finish(self):
        exception = None

        try:
            self.driver.close()
        except Exception as e:
            exception = e

        try:
            self.display.stop()
        except Exception as e:
            exception = e

        if exception is not None:
            raise exception
